src/transformStateWithClones.js

```diff
@@ -7,7 +7,49 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const clones = [];
+
+  for (const { type, extraData, keysToRemove } of actions) {
+    switch (type) {
+      case 'addProperties':
+        addProperties(stateClone, extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateClone, keysToRemove);
+        break;
+
+      case 'clear':
+        clear(stateClone);
+        break;
+
+      default:
+        return;
+    }
+
+    clones.push({ ...stateClone });
+  }
+
+  return clones;
+}
+
+function addProperties(obj, properties) {
+  for (const property in properties) {
+    obj[property] = properties[property];
+  }
+}
```

Comment:
It will be better if you use `Object.assign(target, source)` instead of using a` for...in` loop ✨